src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,454 @@
+ */
+int
+dfs_sys_punch(dfs_sys_t *dfs_sys, const char *file,
+	      daos_off_t offset, daos_off_t len);
+
+/**
+ * Similar to dfs_sys_punch but on an open object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_opunch(dfs_sys_t *dfs_sys, dfs_obj_t *obj,
+	       daos_off_t offset, daos_off_t len);
+
+/**
+ * Remove an object identified by path. If object is a directory and is
+ * non-empty; this will fail unless force option is true. If object is a
+ * symlink, the symlink is removed.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in]	force	If true, remove dir even if non-empty.
+ * \param[in]	oid	Optionally return the DAOS Object ID of the removed obj.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_remove(dfs_sys_t *dfs_sys, const char *path, bool force,
+	       daos_obj_id_t *oid);
+
+/**
+ * Similar to dfs_sys_remove but optionally enforces a type check
+ * on the entry.
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in]	force	If true, remove dir even if non-empty.
+ * \param[in]	mode	mode_t (S_IFREG | S_IFDIR | S_IFLNK).
+ *			Pass 0 skip the type check.
+ * \param[out]	oid	Optionally return the DAOS Object ID of the removed obj.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_remove_type(dfs_sys_t *dfs_sys, const char *path, bool force,
+		    mode_t mode, daos_obj_id_t *oid);
+
+/**
+ * Create a directory.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	dir	Link path of new dir.
+ * \param[in]	mode	mkdir mode.
+ * \param[in]	cid	DAOS object class id (pass 0 for default MAX_RW).
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_mkdir(dfs_sys_t *dfs_sys, const char *dir, mode_t mode,
+	      daos_oclass_id_t cid);
+
+/**
+ * Open a directory.
+ * The directory must be closed with dfs_sys_closedir().
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	dir	Link path of dir.
+ * \param[in]	flags	(O_NOFOLLOW)
+ * \param[out]	dirp	Pointer to the open directory.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_opendir(dfs_sys_t *dfs_sys, const char *dir, int flags, DIR **dirp);
+
+/**
+ * Close a directory opened with dfs_sys_opendir().
+ *
+ * \param[in]	dirp	Pointer to the open directory.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_closedir(DIR *dirp);
+
+/**
+ * Read a directory opened with dfs_sys_opendir().
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	dirp	Pointer to the open directory.
+ * \param[in]	dirent	Pointer to the next directory.
```

Comment:
Needs more explanation, and I suspect dirent is wrong.  I presume at some level there's a need to call this iteratively?

Reply:
Sure I can add more documentation for this.

As for dirent, POSIX readdir returns a `struct dirent *`. Upon reaching the end, it returns NULL.
So, to iterate, you would do something like this:
```
DIR *dirp = NULL;
struct dirent *ent = NULL;
dfs_sys_opendir(dfs_sys, "/some/dir", 0, &dirp);
do {
    dfs_sys_readdir(dfs_sys, dirp, &ent);
} while(ent != NULL)
```
I'll add more documentation for this, but do you think this is okay? It seemed to be the closest approximation to the POSIX equivalent.